### DISCUSSION TITLE
Is NiceGUI Production ready?

### DISCUSSION THREAD

**Discussion by Shahin-rmz** [zauberzeug/nicegui#395] [Q&A] [answered]
Hello,
NeceGUI aims and solves very important problem of Python devs.
It was a day, that Python devs had almost no chance to deliver an app, without doing Javascript.
Then the solutions like `streamlit` and `Dash` came up, which either was not totally free, or does not give the devs enough flexibility.
Then came a solution like `pyscript` to the game, which is not ready yet, and it goes so deep inside front end technology.
You know also `justpy`.

NiceGUI aims the problem correctly and solves the issue in very nice way. it gives enough flexibilty as well as strong and easy way of delivering app.
My question is that, is it production ready?
I play around and see some app demos. some of the demos seems to refresh the page multiple times (I am not sure if it is my browsers problem or a bug from NiceGUI).

does it mean that we should not bring market apps with NiceGUI soon?

I recently discovered another GUI lib, which is called [Pynecone](https://github.com/pynecone-io/pynecone).
there is so much similarities between NiceGUI and Pynecone, as in the eyes of no JS dev.
Would be thankful if someone explains, the difference between NiceGUI and Pynecone.

again, thanks for making Python even more attractive.
Best

**Comment by Diegiwg** [comment]
For me, Nicegui is very stable, I haven't found any errors so far, even doing some 'hacks', to expand the control with Javascript.

Still missing some quality of life features for development, like the styling pattern that is under discussion/implementation #362 #384 #117.

**Comment by Shahin-rmz** [comment]
NiceGUI finds really genius ways for the UI problem.
I will try to make a test app with it and share it with you , so maybe you use it as another example!

**Comment by rodja** [comment, marked as answer]
Thanks for the warm words. We are using NiceGUI in production already. Not only is https://nicegui.io created with NiceGUI itself, but we use it to control market-ready mobile robots and provide tools for our customers.

As far as I can see, the main differences to Pynecone are:

* NiceGUI was initially build for accessing and controlling hardware as shown in our [webcam example](https://github.com/zauberzeug/nicegui/tree/main/examples/opencv_webcam)); I'm not sure how it would be done with Pynecone
* NiceGUI encourages the use of standard Python (callbacks, if-statements,..), Pynecone on the other hand uses explicit State classes and provides constructs like [pc.cond](https://pynecone.io/docs/library/layout/cond) and [pc.foreach](https://pynecone.io/docs/library/layout/foreach).
* NiceGUI uses Vue/Quasar for the frontend while Pynecone is build on NextJS
* NiceGUI generates HTML/JS/CSS via templates on the fly while Pynceone has an explicit compile step; so NiceGUI can be run with normal "Python" instead of using a command like "pc"
* while both frameworks use FastAPI for the backend, in NiceGUI you can actually [use your own App and simply extend it with NiceGUI](https://github.com/zauberzeug/nicegui/tree/main/examples/fastapi) to provide additional UI; Pynecone hides FastAPI which makes it harder to provide other API endpoints (for example to serve images from memory instead of files).

As often, it boils down to personal preferences. While on the surface both frameworks solve the same problem, the architectures and philosophies are quite different.

**Comment by Shahin-rmz** [comment]
Thanks for informative answer. 
many reasons aside, because of nice fastAPI implementation, I love NiceGUI. 
Going to start my first project with it this week.
Best
